src/Nethermind/Nethermind.Synchronization/FastBlocks/SyncStatusList.cs

```diff
@@ -77,6 +79,80 @@ public void GetInfosForBatch(BlockInfo?[] blockInfos)
+        /// </summary>
+        /// <param name="batchSize"></param>
+        /// <param name="blockExist"></param>
+        /// <param name="infos"></param>
+        /// <returns></returns>
+        public bool TryGetInfosForBatch(int batchSize, Func<BlockInfo, bool> blockExist, out BlockInfo?[] infos)
+        {
+            BlockInfo?[] outputArray = new BlockInfo?[batchSize];
+            BlockInfo?[] workingArray = new BlockInfo?[batchSize];
+
+            for (int attempt = 0; attempt < 8; attempt++)
+            {
+                // Because the last clause of GetInfosForBatch increment the _lowestInsertWithoutGap need to be run
+                // sequentially, can't find an easy way to parallelize the checking for block exist part in the check
+                // So here we are...
+                GetInfosForBatch(workingArray);
+
+                bool hasNonNull = false;
+                bool hasInserted = false;
+                Parallel.For(0, workingArray.Length, (i) =>
+                {
+                    if (workingArray[i] is not null)
+                    {
+                        if (blockExist(workingArray[i]))
+                        {
+                            MarkInserted(workingArray[i].BlockNumber);
+                            hasInserted = true;
+                            workingArray[i] = null;
+                        }
+                        else
+                        {
+                            hasNonNull = true;
+                        }
+                    }
+                });
+
+                if (hasNonNull || !hasInserted)
+                {
+                    int slot = 0;
+                    for (int i = 0; i < workingArray.Length; i++)
+                    {
+                        if (workingArray[i] is not null)
+                        {
+                            if (slot < outputArray.Length)
+                            {
+                                outputArray[slot] = workingArray[i];
+                                slot++;
+                            }
+                            else
+                            {
+                                // Not enough space in output we'll need to put back the block
+                                MarkPending(workingArray[i]);
+                            }
+                        }
+                    }
+
+                    infos = outputArray;
+                    return true;
+                }
+
+                // At this point, hasNonNull is false and hasInserted is true, meaning all entry in workingArray
+                // already exist. We switch to a bigger array to improve parallelization throughput
+                if (workingArray.Length < _parallelExistCheckSize)
+                {
+                    workingArray = new BlockInfo[_parallelExistCheckSize];
+                }
+            }
+
+            infos = workingArray;
+            return false;
+        }
```

Comment:
This could benefit from even more comments, or making code more readable in other ways - better naming, or named submethods or something.

Reply:
Separated to inner function